File: pyfcstm/diagnostics/suggested_fix.py
"""Helpers for rendering structured diagnostic suggested fixes."""

from string import Formatter
from typing import Any, Dict, Mapping, Optional, Set

from .codes import CODE_REGISTRY


def _template_field_names(template: str) -> Set[str]:
    out: Set[str] = set()
    for _, field_name, _, _ in Formatter().parse(template):
        if field_name:
            out.add(field_name)
    return out


def render_suggested_fix(
        code: str,
        refs: Mapping[str, Any],
) -> Optional[Dict[str, Any]]:
    """Render the ``codes.yaml`` suggested-fix template for one emit.

    :param code: Diagnostic code.
    :type code: str
    :param refs: Already-rendered diagnostic refs payload.
    :type refs: Mapping[str, Any]
    :return: JSON-friendly suggested-fix payload, or ``None`` if the code
        has no suggested-fix metadata.
    :rtype: Optional[Dict[str, Any]]
    """
    spec = CODE_REGISTRY.get(code)
    if spec is None or spec.suggested_fix is None:
        return None

    values = dict(refs)
    state_path = values.get('state_path')
    if isinstance(state_path, str):
        values.setdefault('state_name', state_path.rsplit('.', 1)[-1])
    composite_path = values.get('composite_path')
    if isinstance(composite_path, str):
        values.setdefault('composite_name', composite_path.rsplit('.', 1)[-1])

    suggested = spec.suggested_fix
    if not suggested.anchor_ref.startswith('refs.'):
        return None
    anchor_field = suggested.anchor_ref[5:]
    if values.get(anchor_field) is None:
        return None
    for field_name in _template_field_names(suggested.text_template):
        if values.get(field_name) is None:
            return None

    text = suggested.text_template.format(**values)
    return {
        'kind': suggested.kind,
        'target': suggested.target,
        'anchor': {
            'type': 'ref',
            'ref': suggested.anchor_ref,
        },
        'text': text,
        'rationale': suggested.rationale,
    }


def refs_with_suggested_fix(
        code: str,
        refs: Mapping[str, Any],
) -> Dict[str, Any]:
    """Return ``refs`` copied with ``suggested_fix`` attached when declared."""
    out = dict(refs)
    fix = render_suggested_fix(code, out)
    if fix is not None:
        out['suggested_fix'] = fix
    return out

File: editors/jsfcstm/src/diagnostics/suggested-fix.ts
import type {CodeRegistry} from './codes';

// Import the JSON payload directly. In ts-node source tests, resolving
// ``./codes`` prefers ``codes.json`` over ``codes.ts``; using the explicit
// filename keeps source and built-package execution aligned.
// eslint-disable-next-line @typescript-eslint/no-var-requires
const codesContents = require('./codes.json') as CodeRegistry;

export interface SuggestedFixPayload {
    kind: 'insert' | 'delete' | 'replace';
    target: string;
    anchor: {
        type: 'ref';
        ref: string;
    };
    text: string;
    rationale: string;
}

function lastPathSegment(value: unknown): string | undefined {
    if (typeof value !== 'string') return undefined;
    const parts = value.split('.');
    return parts[parts.length - 1];
}

function templateFieldNames(template: string): string[] {
    const out: string[] = [];
    const pattern = /\{([A-Za-z_][A-Za-z0-9_]*)\}/g;
    let match = pattern.exec(template);
    while (match !== null) {
        out.push(match[1]);
        match = pattern.exec(template);
    }
    return out;
}

function formatTemplate(template: string, values: Record<string, unknown>): string | null {
    for (const field of templateFieldNames(template)) {
        if (values[field] === undefined || values[field] === null) {
            return null;
        }
    }
    return template.replace(/\{([A-Za-z_][A-Za-z0-9_]*)\}/g, (_, field: string) => String(values[field]));
}

/**
 * Render a ``codes.yaml`` suggested-fix template for one diagnostic emit.
 */
export function renderSuggestedFix(
    code: string,
    refs: Record<string, unknown>,
): SuggestedFixPayload | null {
    const spec = codesContents[code];
    const suggested = spec?.suggested_fix;
    if (!suggested) return null;
    if (!suggested.anchor_ref.startsWith('refs.')) return null;
    const anchorField = suggested.anchor_ref.slice('refs.'.length);
    if (refs[anchorField] === undefined || refs[anchorField] === null) return null;

    const values: Record<string, unknown> = {...refs};
    const stateName = lastPathSegment(values.state_path);
    if (stateName !== undefined && values.state_name === undefined) {
        values.state_name = stateName;
    }
    const compositeName = lastPathSegment(values.composite_path);
    if (compositeName !== undefined && values.composite_name === undefined) {
        values.composite_name = compositeName;
    }

    const text = formatTemplate(suggested.text_template ?? '', values);
    if (text === null) return null;
    return {
        kind: suggested.kind,
        target: suggested.target,
        anchor: {
            type: 'ref',
            ref: suggested.anchor_ref,
        },
        text,
        rationale: suggested.rationale,
    };
}

/**
 * Copy ``refs`` and attach ``suggested_fix`` when code metadata can be
 * rendered safely for the current payload.
 */
export function refsWithSuggestedFix(
    code: string,
    refs: Record<string, unknown>,
): Record<string, unknown> {
    const out: Record<string, unknown> = {...refs};
    const fix = renderSuggestedFix(code, out);
    if (fix) {
        out.suggested_fix = fix;
    }
    return out;
}
